server/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java

```diff
@@ -48,15 +54,35 @@ synchronized Settings getPersistentUpdate() {
         return persistentUpdates.build();
     }
 
-    synchronized ClusterState updateSettings(final ClusterState currentState, Settings transientToApply, Settings persistentToApply) {
+    synchronized ClusterState updateSettings(
+            final ClusterState currentState,
+            final Settings transientToApply,
+            final Settings persistentToApply,
+            final Logger logger) {
         boolean changed = false;
-        Settings.Builder transientSettings = Settings.builder();
-        transientSettings.put(currentState.metaData().transientSettings());
-        changed |= clusterSettings.updateDynamicSettings(transientToApply, transientSettings, transientUpdates, "transient");
 
+        /*
+         * Our cluster state could have unknown or invalid settings that are known and valid in a previous version of Elasticsearch. We can
+         * end up in this situation during a rolling upgrade where the previous version will infect the current version of Elasticsearch
+         * with settings that the current version either no longer knows about or now considers to have invalid values. When the current
+         * version of Elasticsearch becomes infected with a cluster state containing such settings, we need to skip validating such settings
+         * and instead archive them. Consequently, for the current transient and persistent settings in the cluster state we do the
+         * following:
+         *  - create a settings instance with that has archived all unknown or invalid settings
+         *  - split this settings instance into two with the known and valid settings in one, and the unknown or invalid in another
+         *  - validate the incoming settings update combined with the existing known and valid settings
+         *  - merge in the archived unknown or invalid settings
+         */
+        final Tuple<Settings, Settings> partitionedTransientSettings =
```

Comment:
can we have 2 methods `Settings getValidSettings(Settings)` and `Settings getInvalidSetting(Settings)` that would make it simpler to see what is valid and not instead of a tuple?

Reply:
I pushed a change. I kept the tuple but clarified its usage. I hope this helps.